### ISSUE TITLE
Builtin support for decimal types

### ISSUE THREAD

**Issue by jcrist** [jcrist/msgspec#287]
Right now decimal types can be supported using an extension. However, these are common enough we may want to provide builtin support. This won't provide a _huge_ speedboost for encoding/decoding these values, since `decimal.Decimal` objects lack a c-api and are an opaque c-object. However, avoiding dispatching through `enc_hook`/`dec_hook` does result in a ~25% performance increase per my measurements.

Since usage of a `decimal` type implies the user cares about decimal precision, most JSON messages encode these values as strings (`"123.4"`, `"123"`) rather than floats or integers (`123.4`,  `123`). `pydantic` is an exception here - it currently encodes/decodes these values from floats/ints. This means that decimals don't accurately roundtrip, and has been an issue in the past. It looks like in pydantic v2 they're planning on changing these to encode/decode as strings (https://github.com/pydantic/pydantic/issues/1511).

Until we redo our extension support, any strategy chosen for builtin supported types cannot be overridden by the user. I firmly believe strings are the right way to serialize decimal values, but this would mean there'd be no way a user could override support to be compatible with how pydantic V1 treats them. That said, I doubt any users relying on `decimal` support in pydantic would be using the default encoder, since that doesn't properly roundtrip values (negating the reason for using decimals in the first place).

@provinzkraut, @goldziher would making this change be an issue for starlite? This would change how decimal values are encoded/decoded, but (I think) is a better option for these values than what pydantic does.

**Comment by jcrist** [comment]
Also cc @cjdsellers, since y'all are also using `decimal` types with `msgspec`.

**Comment by provinzkraut** [comment]
> would making this change be an issue for starlite? 

It would be somewhat of a breaking change, since we aim to JSON-serialize things the same as `pydantic.BaseModel.json` does, and we currently have a case for this set up in our `enc_hook` that uses [`pydantic.json.decimal_encoder`](https://github.com/pydantic/pydantic/blob/b5f10aa89138e4a589194f3046c593cd183c4d4e/pydantic/json.py#L23).

> but (I think) is a better option for these values than what pydantic does.

I agree. Floats are a really bad choice for decimals, since reason decimals are used is usually because floats aren't appropriate. 

Seeing as this will be a breaking change anyway at some point (because pydantic changes the way decimals are serialized), I guess it's acceptable to make this change a bit earlier on our side as well.

**Comment by jcrist** [comment]
Thanks for the quick feedback y'all! I've implemented this in #288.

> and normally always parse strings to accurately infer precision over the wire.

While what you have is fine, with this change you could rely on `msgspec` to handle the parsing as part of the message, rather than manually calling `decimal.Decimal` yourself:

```python
class BinanceFuturesMarkPriceData(msgspec.Struct, frozen=True):
    ...
    r: decimal.Decimal  # <- will parse the string to a decimal on decode
    ...
```

> Seeing as this will be a breaking change anyway at some point (because pydantic changes the way decimals are serialized), I guess it's acceptable to make this change a bit earlier on our side as well.

When this release comes out, you may also want to change the default encoder in `typed_encoders` (I think that's what starlite calls it?) from `decimal_encoder` to `str` so decimal objects inside pydantic models encode the same as decimal objects outside pydantic models.
